R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+    mz_groups <- mz_groups[keep]
+    mzs <- mzs[keep]
+    ints <- unlist(base::lapply(x, function(z) z@intensity))[mz_order][keep]
+    ## Create result.
+    new_sp <- x[[main]]
+    ## Support also weighted.mean:
+    if (is.character(mzFun) && mzFun == "weighted.mean") {
+        intsp <- split(ints, mz_groups)
+        new_sp@mz <- base::mapply(split(mzs, mz_groups), intsp,
+                                  FUN = function(mz_vals, w)
+                                      stats::weighted.mean(mz_vals, w + 1,
+                                                           na.rm = TRUE),
+                                  USE.NAMES = FALSE, SIMPLIFY = TRUE)
+        new_sp@intensity <- unlist(base::lapply(intsp, intensityFun),
+                                   use.names = FALSE)
+    } else {
+        new_sp@mz <- unlist(base::lapply(split(mzs, mz_groups), mzFun),
+                            use.names = FALSE)
+        new_sp@intensity <- unlist(base::lapply(split(ints, mz_groups),
+                                                intensityFun),
+                                   use.names = FALSE)
+    }
+    if (is.unsorted(new_sp@mz))
+        stop("m/z values of combined spectrum are not ordered")
+    new_sp@peaksCount <- length(new_sp@mz)
+    new_sp
+}
+
+#' Group a sorted `numeric` of m/z values from consecutive scans by ion assuming
+#' that the variation between m/z values for the same ion in consecutive scan
+#' is much lower than the difference between m/z values within one scan.
+#'
+#' @param x `numeric` with ordered and combined m/z values from consecutive
+#'     scans.
+#'
+#' @param mzd `numeric(1)` with the m/z difference below which m/z values are
+#'     grouped together. If not provided the `.estimate_mz_scattering` function
+#'     is used to estimate it.
+#' 
+#' @return `integer` of same length than `x` grouping m/z values.
+#'
+#' @noRd
+.group_mz_values <- function(x, mzd) {
+    mzdiff <- diff(x)
+    if (missing(mzd))
+        mzd <- .estimate_mz_scattering(mzdiff, TRUE)
+    ## Create a vector grouping values with difference in mz values being
+    ## smaller than mzd
+    nvals <- diff(c(0, which(!(mzdiff < mzd)), length(x)))
+    rep(1:length(nvals), nvals)
+}
+
+#' Estimate the extent of random scattering of m/z values of the same ion in
+#' consecutive scans. This bases on the assumption that the random scattering
+#' of m/z values is much smaller than the m/z resolution of the MS instrument.
+#'
+#' Assumes the data is in profile mode.
+#' 
+#' @param x Either values or differences between values.
+#'
+#' @param is_diff `logical(1)` indicating whether `x` are already differences.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.estimate_mz_scattering <- function(x, is_diff = FALSE) {
+    if (!is_diff)
+        x <- diff(x)
+    dens <- density(x, n = max(c(512, length(x) / 2)))
+    ## Find all turning points, i.e. where density is increasing
+    idxs <- which(diff(sign(diff(dens$y))) == 2)
```

Comment:
@sgibb , (don't know why I can't see your note/review here), but you're completely right. it should be `idxs <- which(diff(sign(diff(dens$y))) == 2) + 1`. I'll fix.

Reply:
Done in https://github.com/lgatto/MSnbase/pull/315/commits/c6196769a56748e03a45f4991bb11194803c5216

Reply:
Sorry, just after I hit "Add a comment" I realized that I wanted to "Start a review" and removed the comment just to recreated it with "Start a review". It is now hidden in the review below the "Show outdated" button. Obviously github sends the email notification right away.